clang/labs/lab1/borovkov_sergey/renamedfile.cpp

```diff
@@ -0,0 +1,199 @@
+    }
+    return true;
+  }
+
+  bool VisitDeclRefExpr(clang::DeclRefExpr *expr) {
+    clang::VarDecl *var = clang::dyn_cast<clang::VarDecl>(expr->getDecl());
+    if (type == IdType::Variable && var && var->getName() == cur_name) {
+      rewriter.ReplaceText(expr->getSourceRange(), new_name);
+    }
+    return true;
+  }
+
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (type == IdType::Structure && record->getName() == cur_name) {
+      rewriter.ReplaceText(record->getLocation(), new_name);
+      const auto *destructor = record->getDestructor();
+      if (destructor) {
+        rewriter.ReplaceText(destructor->getLocation(), cur_name.size() + 1,
+                             '~' + new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXConstructorDecl(clang::CXXConstructorDecl *constructor) {
+    if (type == IdType::Structure) {
+      if (constructor->getNameAsString() == cur_name) {
+        rewriter.ReplaceText(constructor->getLocation(), cur_name.size(),
+                             new_name);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXNewExpr(clang::CXXNewExpr *newExpr) {
+    if (type == IdType::Structure) {
+      if (newExpr->getConstructExpr()->getType().getAsString() == cur_name) {
+        rewriter.ReplaceText(newExpr->getExprLoc(), cur_name.size() + 4,
+                             "new " + new_name);
+      }
+    }
+    return true;
+  }
+
+  bool save_changes() { return rewriter.overwriteChangedFiles(); }
+
+private:
+  clang::Rewriter rewriter;
+  IdType type;
+  std::string cur_name;
+  std::string new_name;
+};
+
+class RenameASTConsumer : public clang::ASTConsumer {
+public:
+  explicit RenameASTConsumer(clang::CompilerInstance &CI, IdType type,
+                             clang::StringRef cur_name,
+                             clang::StringRef new_name)
+      : Visitor(clang::Rewriter(CI.getSourceManager(), CI.getLangOpts()), type,
+                cur_name, new_name) {}
+
+  void HandleTranslationUnit(clang::ASTContext &context) override {
+    Visitor.TraverseDecl(context.getTranslationUnitDecl());
+    if (Visitor.save_changes()) {
+      llvm::errs() << "An error occurred while saving changes to a file!\n";
+    }
+  }
+
+private:
+  RenameVisitor Visitor;
+};
+
+class renamedfilePlugin : public clang::PluginASTAction {
+public:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &CI,
+                    clang::StringRef InFile) override {
+    return std::make_unique<RenameASTConsumer>(CI, type, cur_name, new_name);
+  }
+
+protected:
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &args) override {
+    std::vector<std::pair<std::string, std::string>> params = {
+        {"type=", ""}, {"cur-name=", ""}, {"new-name=", ""}};
+
+    if (!args.empty() && args[0] == "help") {
+      PrintHelp(llvm::errs());
+      return true;
```

Comment:
Why `true` is return here?

Reply:
Sorry I was wrong, it should be `false`. If `false` is used instead of `true`, then provided that the first argument is not empty and is equal to "help", the `PrintHelp function(llvm::arc())` will still be called to output help information. However, after that, the `return false`; statement indicates that the operation failed. In the context of this code, this means that if "help" is the first argument, the help information will be printed, but the program will return with an error, which may be useful.